src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.Com.cs

```diff
@@ -207,6 +207,9 @@ public static unsafe void GetNativeVariantForObject(object? obj, IntPtr pDstNati
                 case System.Reflection.Missing value:
                     data->AsError = DISP_E_PARAMNOTFOUND;
                     break;
+                case Enum value:
+                    data->AsI8 = (long)Convert.ToInt32(value);
```

Comment:
Enums are marshalled as their underlying type. For example:
```csharp
using System;
using System.Runtime.InteropServices;

public enum MyEnum : ushort
{
}

[StructLayout(LayoutKind.Sequential)]
internal struct NativeVariant
{
    public ushort vt;
    public ushort wReserved1;
    public ushort wReserved2;
    public ushort wReserved3;

    public IntPtr data1;
    public IntPtr data2;

}  // struct NativeVariant


class Program
{
    static unsafe void Main(string[] args)
    {
        NativeVariant v = default;
        System.Runtime.InteropServices.Marshal.GetNativeVariantForObject((object)default(MyEnum), (IntPtr)(void*)&v);
        Console.WriteLine(v.vt);

    }
}
```
This prints 18 /* VT_UI2 */ on CoreCLR, but I think it will print 20 /* *VT_I8 */ with this change.

Reply:
Another case that this does not support currently are the conversions using `IConvertible`. For example:
```csharp
using System;
using System.Runtime.InteropServices;

public class MyConvertible : IConvertible
{
    public TypeCode GetTypeCode()
    {
        return TypeCode.Int32;
    }

    public bool ToBoolean(IFormatProvider provider)
    {
        throw new NotImplementedException();
    }

    public byte ToByte(IFormatProvider provider)
    {
        throw new NotImplementedException();
    }

    public char ToChar(IFormatProvider provider)
    {
        throw new NotImplementedException();
    }

    public DateTime ToDateTime(IFormatProvider provider)
    {
        throw new NotImplementedException();
    }

    public decimal ToDecimal(IFormatProvider provider)
    {
        throw new NotImplementedException();
    }

    public double ToDouble(IFormatProvider provider)
    {
        throw new NotImplementedException();
    }

    public short ToInt16(IFormatProvider provider)
    {
        throw new NotImplementedException();
    }

    public int ToInt32(IFormatProvider provider)
    {
        return 42;
    }

    public long ToInt64(IFormatProvider provider)
    {
        throw new NotImplementedException();
    }

    public sbyte ToSByte(IFormatProvider provider)
    {
        throw new NotImplementedException();
    }

    public float ToSingle(IFormatProvider provider)
    {
        throw new NotImplementedException();
    }

    public string ToString(IFormatProvider provider)
    {
        throw new NotImplementedException();
    }

    public object ToType(Type conversionType, IFormatProvider provider)
    {
        throw new NotImplementedException();
    }

    public ushort ToUInt16(IFormatProvider provider)
    {
        throw new NotImplementedException();
    }

    public uint ToUInt32(IFormatProvider provider)
    {
        throw new NotImplementedException();
    }

    public ulong ToUInt64(IFormatProvider provider)
    {
        throw new NotImplementedException();
    }
}

[StructLayout(LayoutKind.Sequential)]
internal struct NativeVariant
{
    public ushort vt;
    public ushort wReserved1;
    public ushort wReserved2;
    public ushort wReserved3;

    public IntPtr data1;
    public IntPtr data2;

}  // struct NativeVariant


class Program
{
    static unsafe void Main(string[] args)
    {
        NativeVariant v = default;
        System.Runtime.InteropServices.Marshal.GetNativeVariantForObject(new MyConvertible(), (IntPtr)(void*)&v);
        Console.WriteLine(v.vt);
    }
}
```
This will print 3 /* VT_I4 */ on CoreCLR. Note that enums implement `IConvertible`, so if you implement the more general IConvertible case, enums should just fall out naturally.

Reply:
Question. Does this warrant having tests somewhere? I test my first implementation on tests from Interop/PInvoke/Variant and nothing was failed. So I naively assume that everything was ok. I would go ahead and implement them, but I'm not sure that you care about this kind of behaviour.

Reply:
`Marshal.GetObjectForNativeVariant` is API. API tests live under src/libraries. The tests for this one are in https://github.com/dotnet/runtime/blob/main/src/libraries/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/GetNativeVariantForObjectTests.cs. I can see that it includes the tests for enums.

Reply:
Another easy case to handle is throwing ArgumentException for CriticalHandle and SafeHandle